Shipping labels: Address verification

0nko opened the pull request:
Fixes #1716.

This PR adds client support for the address verification endpoint.

**To test:**

1. Authenticate
2. Go to Woo -> Shipping labels
3. Select site with shipping labels
4. Tap on the Verify address button
5. Fill out the form with a valid address
6. Tap on the Verify address button
7. Notice a normalized address object is returned
8. Fill out an invalid address
9. Notice an error is displayed

0nko committed:
Add address verification REST client and store methods

0nko committed:
Add address verification fragment

0nko committed:
Add address verification button to labels fragment

0nko committed:
Update shipping address model

0nko committed:
Add the address verification endpoint address

0nko committed:
Update the model properties

0nko committed:
Add a helper extension function for generating maps from objects

0nko committed:
Change the GET request to POST

0nko committed:
Fix the parameter serialization

0nko committed:
Remove unused code

0nko committed:
Add radio buttons for origin & destination

0nko committed:
Update REST client and store result type

0nko committed:
Display the verification result in the log

0nko committed:
Optimize imports

0nko committed:
Fix ktlint issues

anitaa1990 commented on fluxc/src/main/java/org/wordpress/android/fluxc/network/utils/JsonExt.kt:
```diff
@@ -0,0 +1,16 @@
+import com.google.gson.reflect.TypeToken
+
+// Convert an object of type T to type R
+inline fun <T, reified R> T.convert(): R {
```
Nice <3

anitaa1990 reviewed (commented):
Hey @0nko! The code changes look good! I just tfaced a couple of issues when testing: 

- It look like the country field is mandatory for the API request. I am getting a 400 error when the country is blank. Can we log an error message here to make sure atleast country is entered before calling the API?

- I am getting this crash when entering the country that is not the US.
```
E: FATAL EXCEPTION: main
    Process: org.wordpress.android.fluxc.example, PID: 5907
    java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String org.wordpress.android.fluxc.network.rest.wpcom.wc.WooError.getMessage()' on a null object reference
        at org.wordpress.android.fluxc.store.WCShippingLabelStore$verifyAddress$2.invokeSuspend(WCShippingLabelStore.kt:114)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
```

I am not actually able to test the error scenario because when I enter a dummy address, the app is crashing with the above error.

- Could we add some unit tests to `WCShippingLabelStoreTest`?

0nko committed:
Split the error types

0nko committed:
Add message if there is none

0nko committed:
Add country validation

0nko committed:
Make the result type a data class for easier comparison

0nko committed:
Add some simple unit tests

0nko committed:
Fix ktlint issue

0nko commented:
Thanks @anitaa1990!

> It look like the country field is mandatory for the API request. I am getting a 400 error when the country is blank. Can we log an error message here to make sure atleast country is entered before calling the API?

Done.

> I am getting this crash when entering the country that is not the US.

This is strange because I'm able to test it with different countries. I added some checks that should prevent the crash, though.

> Could we add some unit tests to WCShippingLabelStoreTest?

Fixed.

Ready for another round.

anitaa1990 reviewed (approved):
Looks good! Nice work @0nko :shipit:
